### PR TITLE
Clinton/der model

### PR DESCRIPTION
**Pull request by clinton0313**
Model is mostly correct, but I am not a hundred percent certain. Couple points:

1) There are methods I have implemented that are consistent with the methods in the original paper's published github code (in particular the DeepEvidentialCallback), but these methods are suspiciously not mentioned in the paper: https://github.com/aamini/evidential-deep-learning

2) I have tried to reproduce the exact result of the cubic test as in the above github, but even while setting the same seeds I cannot get it exactly, and some of the variance around the training points are not as smooth so there is some uncertainty whether this is correct. Caveat is that I am able to produce many results that are similar in some way, but from seed to seed or initialization to initialization the results can vary wildly. A little put off though, since I have cloned amini's repo and realized that the perfect figure that he reports is only perfect for seed(1234) and if you change the seeds for his code, you will also get wildly different results of the model. Seems like he cherry picked his results. 

3) One avenue to potentially explore afterwards is maybe adding dropout and batch_normalization between layers to see if this can help regularize the results? 

4) There are two extra files that will be in the future deleted or moved out of the models folder: amini.pkl and model_test.py which is what I have been using to diagnose the model. (see final commit note).

### CODE REVIEW

File: trieste/models/interfaces.py [hunk 1]
```diff
@@ -300,6 +300,31 @@ def sample_ensemble(self, query_points: TensorType, num_samples: int) -> TensorT
         """
         raise NotImplementedError
 
+@runtime_checkable
+class EvidentialPriorModel(ProbabilisticModel, Protocol):
+    """
+    This is an interface for models that have a normal prior distribution over the outputs 
+    defined by a Normal Inverse Gamma higher order evidential distribution. these models can act
+    as probabilistic models by deriving estimates of epistemic uncertainty from the variance of mu
+    and aleatoric uncertainty from the expectation of sigma if it has a higher order distribution.
```
Comment (ampudia19) [inline trieste/models/interfaces.py]: Assuming the prior normal, we will always have at least two moments, which potentially makes the last clarification comment a bit redundant and confusing.

File: trieste/models/keras/architectures.py [hunk 2]
```diff
@@ -233,3 +235,113 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
...
+
+        output_layer = tf.keras.layers.Dense(
+            units= 4 * self.flattened_output_shape,
+            dtype=self.input_tensor_spec.dtype,
+            name="output"
+        )
+
+        return output_layer
+
+    def _get_evidence(self, evidence: tf.Tensor) -> tf.Tensor:
```
Comment (ampudia19) [inline trieste/models/keras/architectures.py:321]: This of course works, but I don't quite fully see the point of making _get_evidence be Tensor-specific, and not an array of three tensors where you then apply the transformations as in below.

File: trieste/models/keras/builders.py [hunk 3]
```diff
@@ -80,3 +80,38 @@ def build_vanilla_keras_ensemble(
     keras_ensemble = KerasEnsemble(networks)
 
     return keras_ensemble
+
+
+def build_vanilla_keras_deep_evidential(
+    data: Dataset,
+    num_hidden_layers: int = 3,
+    units: int = 100,
+    activation: Union[str, tf.keras.layers.Activation] = "relu",
```
Comment (ampudia19) [inline trieste/models/keras/builders.py:89]: I'm guessing you have access to the cloned repo, but I do wonder whether they use softplus for all layers, considering it approximates the ReLu.

File: trieste/models/keras/utils.py [hunk 4]
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
...
+        that characterize the normal inverse gamma distribution given in the order: gamma,
+        lambda, alpha, beta.
+    :return: The loss values
+    '''
+
+    gamma, v, alpha, beta = tf.split(y_pred, 4, axis=-1)
+
+    omega = 2 * beta * (1 + v)
+
+    negative_log_likelihood = (
```
Comment (ampudia19) [inline trieste/models/keras/utils.py:121]: There's a naming conflict here, deep ensembles uses the negative_log_likelihood function. Not an issue, but perhaps we should avoid any naming overlap.

File: trieste/models/keras/architectures.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
 import tensorflow as tf
 import tensorflow_probability as tfp
 
+from trieste.models.keras.utils import normal_inverse_gamma_negative_log_likelihood, normal_inverse_gamma_regularizer
```
Comment (ampudia19) [inline trieste/models/keras/architectures.py:28]: These are unused as you correctly compile the model in models.

File: trieste/models/keras/utils.py [hunk 6]
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
...
+    :param y_true: The output variable values.
+    :param y_pred: The four output parameters of the deep evidential regression model
+        that characterize the normal inverse gamma distribution given in the order: gamma,
+        lambda, alpha, beta.
+    :return: The loss values
+    '''
+    gamma, v, alpha, _ = tf.split(y_pred, 4, axis=-1)
+    return tf.reduce_mean(tf.abs(y_true - gamma) * (2*v + alpha), axis=0)
+
+def deep_evidential_regression_loss(
```
Comment (ampudia19) [inline trieste/models/keras/utils.py]: This one is useless now, right? You use both losses in a list when compiling, and use the callback to calibrate lambda. To the best of my ability, I can't think nor see what use you give this function now.

File: trieste/models/keras/utils.py [hunk 7]
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
...
+    return loss_fn
+
+class DeepEvidentialCallback(tf.keras.callbacks.Callback):
+    def __init__(self, reg_weight, maxi_rate, epsilon, verbose):
+        self.reg_weight = reg_weight
+        self.maxi_rate = maxi_rate
+        self.epsilon = epsilon
+        self.verbose = verbose
+
+    def on_batch_end(self, epoch, logs=None):
```
Comment (ampudia19) [inline trieste/models/keras/utils.py]: on_batch_end has no epoch argument, and in any case you don't use it.

File: trieste/models/keras/utils.py [hunk 8]
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
...
+        self.reg_weight.assign_add(self.maxi_rate * (logs["output_2_loss"] - self.epsilon))
```
Comment (ampudia19) [inline trieste/models/keras/utils.py]: I can't find this anywhere in the paper, but I recall you saying this comes from his paper code. Are maxi_rates and epsilons calibrated somehow, or are these tunable hyperparams?

File: trieste/models/keras/sampler.py [hunk 9]
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
...
+        """
+        Call trajectory function. Makes a draw from the posterior normal distribution
+        of outputs given the predicted evidential parameters. 
+        """
+
+        tf.debugging.assert_equal(
+            self._model.model.flattened_output_shape,
+            tf.shape(x)[-1],
+            message=f"""
+            IS THIS EVEN THE RIGHT TEST??
```
Comment (ampudia19) [inline trieste/models/keras/sampler.py]: In the newer samplers they only check that the batch dimension is correct, a bit tautological since we set it so two lines above.

File: trieste/models/keras/sampler.py [hunk 10]
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
...
+            tf.shape(x)[-2],
+            self._batch_size.value(),
+            message=f"""
+            This trajectory only supports batch sizes of {self._batch_size}.
+            If you wish to change the batch size you must get a new trajectory
+            by calling the get_trajectory method of the trajectory sampler.
+            """
+        )
+
+        dist = tfp.distributions.Normal(self._mu, self._sigma)
```
Comment (ampudia19) [inline trieste/models/keras/sampler.py]: These are not deterministic, but the solution here is straightforward and potentially compatible with graph execution (to use seeds in the tfp.distributions.Normal). I'll happily work on this once code is merged.

File: trieste/models/keras/sampler.py [hunk 11]
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
...
+
+    def resample(self) -> None:
+        """
+        Efficiently resample in-place without retracing. By redrawing mu and sigma vectors based off
+        of already saved evidential parameters.
+        """
+        tf.debugging.Assert(self._evidential_parameters is not None, self._evidential_parameters, \
+            "Trajectory sampler needs to have been called ot initialize before resampling.")
+        gamma, lamb, alpha, beta = tf.split(self._evidential_parameters, 4, axis = -1)
+        self._mu, self._sigma = self._model.sample_normal_parameters(gamma, lamb, alpha, beta, 1)
```
Comment (ampudia19) [inline trieste/models/keras/sampler.py]: a possibly more computationally friendly way is to reset the seeds that you then use to draw from the tfp.distributions.Normal, as per my comment above.

File: trieste/models/keras/models.py [hunk 12]
```diff
@@ -334,25 +343,159 @@ def optimize(self, dataset: Dataset) -> None:
...
+        """
+
+        super().__init__(optimizer)
+
+        if not self.optimizer.fit_args:
+            self.optimizer.fit_args = {
+                "verbose": 0,
+                "epochs": 1000,
+                "batch_size": 32,
+                "callbacks": [
```
Comment (ampudia19) [inline trieste/models/keras/models.py:428]: It is potentially cleaner to include by default the DEC callback, and check (for example given the repr dunder) whether the list of custom callbacks passed by the user includes DEC, and add it otherwise.

File: trieste/models/keras/models.py [hunk 13]
```diff
@@ -334,25 +343,159 @@ def optimize(self, dataset: Dataset) -> None:
...
     def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
-        ...
-    
+
+        evidential_output = self.model(query_points)[0]
+        gamma, v, alpha, beta = tf.split(evidential_output, 4, axis=-1)
+
+        mu, sigma = self.sample_normal_parameters(gamma, v, alpha, beta, num_samples)
+       
+        observation_dist = tfp.distributions.Normal(mu, sigma**0.5)
```
Comment (ampudia19) [inline trieste/models/keras/models.py:530]: Just to confirm, this observation_dist is a N-dimensional multivariate gaussian, just like a GP, and your sigma are the diagonal elements of the covar matrix.

File: trieste/models/keras/architectures.py [hunk 14]
```diff
@@ -233,3 +235,113 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
...
+        output = self.output_layer(hidden_output)
+        
+        gamma, v, alpha, beta = tf.split(output, 4, axis=-1)
+        v = self._get_evidence(v)
+        beta = self._get_evidence(beta)
+        # We need alpha > 1 so that the aleatoric and epistemic uncertainties will
+        # be positive. 
+        alpha = self._get_evidence(alpha) + 1
+        output = tf.concat([gamma, v, alpha, beta], axis=-1)
+        return output, output
```
Comment (ampudia19) [inline trieste/models/keras/architectures.py:358]: Docstrings should mention this, as it may otherwise be confusing and hard to grasp.

**Review by ampudia19 — COMMENTED** [review]
All tests run for me, and the comments I left are clarifications or very minor adjustments. Overall, a very clean implementation!

In regards to trying to further approximate Amini's original results, I have tried to create shallower models that may overfit less on the populated space, but never got his level of smoothness. Seeing the output of your attempts at adding dropout and batch_normalization, the latter seems to cause some trouble for predictions around zero so I'd ignore it. Using dropout does not seem to harm, and you could push it as well.

**Comment by ampudia19** [comment]
Once this is pushed, we should try to merge in the code changes from hrvoje's de_sampler branch

File: trieste/models/keras/builders.py (hunk 3 above)
Comment (clinton0313) [reply trieste/models/keras/builders.py:89]: No he does not. He uses a relu. I once accidentally put in softplus for the activation and the results get a little crazy.

Comment (ampudia19) [reply trieste/models/keras/builders.py:89]: Indeed, while testing it the results were all over the place. Seems like a bit of a theme with DER.

File: trieste/models/keras/utils.py (hunk 8 above)
Comment (clinton0313) [reply trieste/models/keras/utils.py]: It is not in the paper just in his code. They seem to just be hyperparameters. 

Here is the code for the paper's trainer: 
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/trainers/evidential.py

The defaults are lambda = 0 (reg_weight), maxi_rate = 1e-4, epsilon = 1e-2. 

Here for the cubic tests he is using lambda = 1e-2 (reg_weight) and maxi_rate = 0:
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/run_cubic_tests.py

Here he is instantiating the trainer with its defaults (line 47) to run the benchmark datasets:
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/run_uci_dataset_tests.py

Comment (ampudia19) [reply trieste/models/keras/utils.py]: Have you tried saving the best model with a callback, like he does?

File: trieste/models/keras/architectures.py (hunk 2 above)
Comment (clinton0313) [reply trieste/models/keras/architectures.py:321]: Just seemed cleaner than split, apply, slice, +1, concat, return

File: trieste/models/keras/models.py (hunk 13 above)
Comment (clinton0313) [reply trieste/models/keras/models.py:530]: Well I guess it technically is that, but really this function instantiates n different normal distributions for each of n data points. mu, and sigma would be nx1 vectors.

File: trieste/models/keras/sampler.py (hunk 10 above)
Comment (clinton0313) [reply trieste/models/keras/sampler.py]: I pretty much just stopped working on this as soon as I realized I didn't understand the way the dimensions were working and waited for Hrvoje's response to my questions. Haven't touched the sampler since. Still a bit confusing to me at first thought whether it is the prediction which would just be mu or the distribution that constitutes the trajectory

Comment (ampudia19) [reply trieste/models/keras/sampler.py]: In the case of ensembles and you can do either (notice he has this "use_samples" argument to return the distribution and sample from it, and otherwise the default behavior is the predictions from networks in the ensemble). Since I'm also stuck at the sampler, let's just see what Hrvoje has to say, and (if you don't mind) I'll take care of it.
